ipc/provider/src/checkpoint.rs

```diff
@@ -162,30 +178,71 @@ impl<T: BottomUpCheckpointRelayer + Send + Sync + 'static> BottomUpCheckpointMan
                     .await?;
                 log::debug!("bottom up bundle: {bundle:?}");
 
-                let epoch = self
-                    .parent_handler
-                    .submit_checkpoint(
-                        submitter,
-                        bundle.checkpoint,
-                        bundle.signatures,
-                        bundle.signatories,
-                    )
+                // We support parallel checkpoint submission using FIFO order with a limited parallelism (controlled by
+                // the size of submission_semaphore).
+                // We need to acquire a permit (from a limited permit pool) before submitting a checkpoint.
+                // We may wait here until a permit is available.
+                let parent_handler_clone = Arc::clone(&self.parent_handler);
+                let submission_permit = self
+                    .submission_semaphore
+                    .clone()
+                    .acquire_owned()
                     .await
-                    .map_err(|e| {
-                        anyhow!(
-                            "cannot submit bottom up checkpoint at height {} due to: {e:}",
-                            event.height
-                        )
-                    })?;
-
-                log::info!(
-                    "submitted bottom up checkpoint({}) in parent at height {}",
-                    event.height,
-                    epoch
+                    .unwrap();
+                all_submit_tasks.push(tokio::task::spawn(async move {
+                    match Self::submit_checkpoint(parent_handler_clone, submitter, bundle, event)
+                        .await
+                    {
+                        Ok(()) => {
+                            log::debug!("Successfully submitted checkpoint");
+                        }
+                        Err(_err) => {
+                            log::debug!("Failed to submit checkpoint");
```

Comment:
```suggestion
                        Err(err) => {
                            log::error!(error = err, height = bundle.height, "Failed to submit checkpoint");
```
Not sure how to log the height, but ignoring errors and even demoting the log to `debug` level is a recipe for frustration.

Reply:
The `error = err`, `height = bundle.height` do not seem to work. I just log them in a string instead.

Reply:
@aakoshh key values were added to the `log` crate in v0.4.21, but our `Cargo.lock` is fixed on v0.4.20. We'll need to upgrade before we can adopt them without using unlocking the `unstable_kv` feature.